--- a/usr/share/perl5/Lemonldap/NG/Manager/Build/Attributes.pm
+++ b/usr/share/perl5/Lemonldap/NG/Manager/Build/Attributes.pm
@@ -1951,6 +1951,7 @@ sub attributes {
                 { k => '::Redis', v => 'Redis' },
                 { k => '::Pg',    v => 'PostgreSQL' },
                 { k => '::MQTT',  v => 'MQTT' },
+                { k => '::Web',   v => 'Web' },
             ],
             flags => 'hp',
         },
