--- a/usr/share/perl5/Lemonldap/NG/Common.pm
+++ b/usr/share/perl5/Lemonldap/NG/Common.pm
@@ -1,6 +1,6 @@
 package Lemonldap::NG::Common;
 
-our $VERSION = '2.21.0';
+our $VERSION = '2.21.1';
 
 1;
 __END__
--- a/usr/share/lemonldap-ng/bin/encryptTotpSecrets
+++ b/usr/share/lemonldap-ng/bin/encryptTotpSecrets
@@ -24,6 +24,8 @@ GetOptions(
 eval {
     POSIX::setgid( scalar( getgrnam('www-data') ) );
     POSIX::setuid( scalar( getpwnam('www-data') ) );
+    my (undef, undef, undef, undef, undef, undef, undef, $homedir, undef) = getpwnam('www-data');
+    $ENV{HOME} = $homedir if $homedir;
 };
 
 sub verbose {
--- a/usr/share/lemonldap-ng/bin/lemonldap-ng-sessions
+++ b/usr/share/lemonldap-ng/bin/lemonldap-ng-sessions
@@ -42,6 +42,8 @@ eval {
     no warnings;
     POSIX::setgid( scalar( getgrnam($opt_group) ) );
     POSIX::setuid( scalar( getpwnam($opt_user) ) );
+    my (undef, undef, undef, undef, undef, undef, undef, $homedir, undef) = getpwnam($opt_user);
+    $ENV{HOME} = $homedir if $homedir;
 };
 
 my $action = shift @ARGV;
--- a/usr/share/lemonldap-ng/bin/rotateOidcKeys
+++ b/usr/share/lemonldap-ng/bin/rotateOidcKeys
@@ -35,6 +35,8 @@ eval {
     no warnings;
     POSIX::setgid( scalar( getgrnam($opt_group) ) );
     POSIX::setuid( scalar( getpwnam($opt_user) ) );
+    my (undef, undef, undef, undef, undef, undef, undef, $homedir, undef) = getpwnam($opt_user);
+    $ENV{HOME} = $homedir if $homedir;
 };
 
 my $debug = $opts->{debug};
