--- a/usr/share/perl5/Lemonldap/NG/Common/Conf/Backends/_DBI.pm
+++ b/usr/share/perl5/Lemonldap/NG/Common/Conf/Backends/_DBI.pm
@@ -56,7 +56,7 @@ sub lastCfg {
 sub _dbh {
     my $self = shift;
     $self->{dbiTable} ||= "lmConfig";
-    return $self->{_dbh} if ( $self->{_dbh} and $self->{_dbh}->ping );
+    return $self->{_dbh} if ( eval { $self->{_dbh} and $self->{_dbh}->ping } );
     eval {
         $self->{_dbh} =
           DBI->connect_cached( $self->{dbiChain}, $self->{dbiUser},
