--- a/usr/share/perl5/Lemonldap/NG/Manager/Attributes.pm
+++ b/usr/share/perl5/Lemonldap/NG/Manager/Attributes.pm
@@ -2445,7 +2445,7 @@ sub attributes {
                                                  'type' => 'select'
                                                },
           'oidcRPMetaDataOptionsLogoutUrl' => {
-                                                'type' => 'url'
+                                                'type' => 'text'
                                               },
           'oidcRPMetaDataOptionsOfflineSessionExpiration' => {
                                                                'type' => 'int'
