--- a/usr/share/perl5/Lemonldap/NG/Manager/Build/Attributes.pm
+++ b/usr/share/perl5/Lemonldap/NG/Manager/Build/Attributes.pm
@@ -4942,6 +4942,10 @@ m{^(?:ldapi://[^/]*/?|\w[\w\-\.]*(?::\d{1,5})?|ldap(?:s|\+tls)?://\w[\w\-\.]*(?:
             type          => 'bool',
             documentation => 'Drop CORS headers from OIDC issuer responses',
         },
+        oidcServiceMetadataTtl => {
+            type          => 'int',
+            documentation => 'OIDC Metadata TTL',
+        },
 
         # OpenID Connect metadata nodes
         oidcOPMetaDataNodes => {
--- a/usr/share/perl5/Lemonldap/NG/Manager/Build/Tree.pm
+++ b/usr/share/perl5/Lemonldap/NG/Manager/Build/Tree.pm
@@ -1583,6 +1583,7 @@ sub tree {
                         'oidcServiceIDTokenExpiration',
                         'oidcServiceAccessTokenExpiration',
                         'oidcServiceOfflineSessionExpiration',
+                        'oidcServiceMetadataTtl',
                     ]
                 },
                 {
