diff --git a/usr/share/perl5/Lemonldap/NG/Common/Conf/Backends/File.pm b/usr/share/perl5/Lemonldap/NG/Common/Conf/Backends/File.pm
index 83ac49ca2..275fcf769 100644
--- a/usr/share/perl5/Lemonldap/NG/Common/Conf/Backends/File.pm
+++ b/usr/share/perl5/Lemonldap/NG/Common/Conf/Backends/File.pm
@@ -123,6 +123,7 @@ sub load {
         }
         binmode FILE;
         $f = join( '', <FILE> );
+        close FILE;
         eval { $ret = from_json( $f, { allow_nonref => 1 } ) };
         if ($@) {
             print STDERR "$@\n";
