--- a/usr/share/perl5/Lemonldap/NG/Common/PSGI.pm
+++ b/usr/share/perl5/Lemonldap/NG/Common/PSGI.pm
@@ -449,17 +449,16 @@ sub logAndRun {
     # register the request object to the logging system
     if ( ref( $self->logger ) and $self->logger->can('setRequestObj') ) {
         $self->logger->setRequestObj($req);
+        $self->logger->info( "New request "
+              . ref($self) . " "
+              . $req->method . " "
+              . $req->request_uri );
     }
     if ( ref( $self->userLogger ) and $self->userLogger->can('setRequestObj') )
     {
         $self->userLogger->setRequestObj($req);
     }
 
-    $self->logger->info( "New request "
-          . ref($self) . " "
-          . $req->method . " "
-          . $req->request_uri );
-
     my $res = $sub->($req);
 
     # Clear the logging system before the next request
