--- a/usr/share/perl5/Lemonldap/NG/Common/Logger/Syslog.pm
+++ b/usr/share/perl5/Lemonldap/NG/Common/Logger/Syslog.pm
@@ -1,7 +1,7 @@
 package Lemonldap::NG::Common::Logger::Syslog;
 
 use strict;
-use Sys::Syslog qw(:standard);
+use Sys::Syslog qw(:standard :extended);
 
 our $VERSION = '2.0.15';
 
@@ -17,6 +17,9 @@ sub new {
         $self->{facility} = $conf->{syslogFacility} || 'daemon';
         $self->{options}  = $conf->{syslogOptions}  || 'cons,pid,ndelay';
     }
+    if ( $conf->{syslogSockOptions} ) {
+        setlogsock( $conf->{syslogSockOptions} );
+    }
     eval { openlog( 'LLNG', $self->{options}, $self->{facility} ) };
     no warnings 'redefine';
     my $show = 1;
