--- a/usr/share/perl5/Lemonldap/NG/Portal/Issuer/OpenIDConnect.pm
+++ b/usr/share/perl5/Lemonldap/NG/Portal/Issuer/OpenIDConnect.pm
@@ -38,7 +38,11 @@ with 'Lemonldap::NG::Portal::Lib::LazyLoadedConfiguration',
 # INTERFACE
 
 sub beforeAuth { 'exportRequestParameters' }
-use constant hook => { updateSessionId => 'updateOidcSecondarySessions' };
+use constant hook => {
+    updateSessionId => 'updateOidcSecondarySessions',
+    eventLogout     => 'rtLogout',
+    unAuthLogout    => 'rtLogout'
+};
 
 # INITIALIZATION
 
@@ -2636,90 +2640,8 @@ sub logout {
 
                 # BACK CHANNEL
                 elsif ( $rpConf->{oidcRPMetaDataOptionsLogoutType} eq 'back' ) {
-
-                    # Logout token must contain:
-                    #  - iss: issuer identifier
-                    #  - sub: subject id (user id)
-                    #  OR/AND
-                    #  - sid: OP session id given to the RP
-                    #  - aud: audience
-                    #  - iat: issue at time
-                    #  - jti: JWT token id
-                    #  - events: should be :
-                    #   { 'http://schemas.openid.net/event/backchannel-logout"
-                    #     => {} } # or a JSON object
-                    #
-                    # Logout token should be send using a POST request:
-                    #
-                    #   POST /backChannelUri HTTP/1.1
-                    #   Host: rp
-                    #   Content-Type: application/x-www-form-urlencoded
-                    #
-                    #   logout_token=<JWT value>
-                    #
-                    # RP response should be 200 (204 accepted) or 400 for errors
-                    my $alg =
-                      $self->rpOptions->{$rp}
-                      ->{oidcRPMetaDataOptionsAccessTokenSignAlg}
-                      || (
-                        $self->conf->{oidcServiceKeyTypeSig} eq 'EC'
-                        ? 'ES256'
-                        : 'RS256'
-                      );
-                    $self->logger->debug(
-                        "Access Token signature algorithm: $alg");
-                    my $userId =
-                      $self->getUserIDForRP( $req, $rp, $req->userData );
-                    my $logoutToken = {
-                        iss => $self->get_issuer($req),
-                        sub => $userId,
-                        aud => $self->getAudiences($rp),
-                        iat => time,
-
-                        # Random string: no response expected from RP
-                        jti => join( "",
-                            map { [ "0" .. "9", 'A' .. 'Z' ]->[ rand 36 ] }
-                              1 .. 8 ),
-                        events => { $self->BACKCHANNEL_EVENTSKEY => {} },
-                    };
-                    if ( $self->rpOptions->{$rp}
-                        ->{oidcRPMetaDataOptionsLogoutSessionRequired} )
-                    {
-                        $logoutToken->{sid} =
-                          $self->getSidFromSession( $rp, $req->{sessionInfo} );
-                    }
-                    $self->logger->debug( "Logout token content: "
-                          . JSON::to_json($logoutToken) );
-                    my $jwt = $self->encryptToken(
-                        $rp,
-                        $self->createJWT( $logoutToken, $alg, $rp ),
-                        $self->rpOptions->{$rp}
-                          ->{oidcRPMetaDataOptionsLogoutEncKeyMgtAlg},
-                        $self->rpOptions->{$rp}
-                          ->{oidcRPMetaDataOptionsLogoutEncContentEncAlg},
-                    );
-                    my $resp = $self->ua->post(
-                        $url,
-                        { logout_token => $jwt },
-                        'Content-Type' => 'application/x-www-form-urlencoded',
-                    );
-                    if ( $resp->is_error ) {
-                        $self->logger->warn(
-                                "OIDC back channel: unable to unlog"
-                              . " $userId from $rp: "
-                              . $resp->message );
-                        $self->logger->debug("Logout token: $jwt");
-                        $self->logger->debug(
-                            'Upstream status: ' . $resp->status_line );
-                        $self->logger->debug(
-                            'Upstream response: ' . ( $resp->content // '' ) );
-                        $code = PE_SLO_ERROR;
-                    }
-                    else {
-                        $self->logger->info(
-                            "OIDC back channel: user $userId unlogged from $rp"
-                        );
-                    }
+                    my $c = $self->_backChannelLogout( $req, $rp, $req->userData );
+                    $code = $c if $c;
                 }
             }
         }
@@ -2728,6 +2650,77 @@ sub logout {
 }
 
 # Internal methods
+sub _backChannelLogout {
+    my ( $self, $req, $rp, $sessionInfo ) = @_;
+
+    # Logout token must contain:
+    #  - iss: issuer identifier
+    #  - sub: subject id (user id)
+    #  OR/AND
+    #  - sid: OP session id given to the RP
+    #  - aud: audience
+    #  - iat: issue at time
+    #  - jti: JWT token id
+    #  - events: should be :
+    #   { 'http://schemas.openid.net/event/backchannel-logout"
+    #     => {} } # or a JSON object
+    #
+    # Logout token should be send using a POST request:
+    #
+    #   POST /backChannelUri HTTP/1.1
+    #   Host: rp
+    #   Content-Type: application/x-www-form-urlencoded
+    #
+    #   logout_token=<JWT value>
+    #
+    # RP response should be 200 (204 accepted) or 400 for errors
+    my $alg         = $self->getSignAlg($rp);
+    my $userId      = $self->getUserIDForRP( $req, $rp, $sessionInfo );
+    my $logoutToken = {
+        iss => $self->get_issuer($req),
+        sub => $userId,
+        aud => $self->getAudiences($rp),
+        iat => time,
+
+        # Random string: no response expected from RP
+        jti =>
+          join( "", map { [ "0" .. "9", 'A' .. 'Z' ]->[ rand 36 ] } 1 .. 8 ),
+        events => { $self->BACKCHANNEL_EVENTSKEY => {} },
+    };
+    if ( $self->rpOptions->{$rp}->{oidcRPMetaDataOptionsLogoutSessionRequired} )
+    {
+        $logoutToken->{sid} =
+          $self->getSidFromSession( $rp, $sessionInfo );
+    }
+    $self->logger->debug(
+        "Logout token content: " . JSON::to_json($logoutToken) );
+    my $jwt = $self->encryptToken(
+        $rp,
+        $self->createJWT( $logoutToken, $alg, $rp ),
+        $self->rpOptions->{$rp}->{oidcRPMetaDataOptionsLogoutEncKeyMgtAlg},
+        $self->rpOptions->{$rp}->{oidcRPMetaDataOptionsLogoutEncContentEncAlg},
+    );
+    my $resp = $self->ua->post(
+        $self->rpOptions->{$rp}->{oidcRPMetaDataOptionsLogoutUrl},
+        { logout_token => $jwt },
+        'Content-Type' => 'application/x-www-form-urlencoded',
+    );
+    if ( $resp->is_error ) {
+        $self->logger->warn( "OIDC back channel: unable to unlog"
+              . " $userId from $rp: "
+              . $resp->message );
+        $self->logger->debug("Logout token: $jwt");
+        $self->logger->debug( 'Upstream status: ' . $resp->status_line );
+        $self->logger->debug(
+            'Upstream response: ' . ( $resp->content // '' ) );
+        return PE_SLO_ERROR;
+    }
+    else {
+        $self->logger->info(
+            "OIDC back channel: user $userId unlogged from $rp");
+    }
+    return PE_OK;
+}
 
 sub metadata {
     my ( $self, $req ) = @_;
@@ -3100,4 +3093,29 @@ sub updateOidcSecondarySessions {
     return;
 }
 
+sub rtLogout {
+    my ( $self, $req, $msg ) = @_;
+
+    # $msg is set when the logout comes from an event
+    my $id;
+    unless ( $msg and $id = $msg->{id} ) {
+        my $auth = $req->{env}->{HTTP_AUTHORIZATION};
+        if ( $auth and my $rt = ( $auth =~ /^Bearer (.+)$/i )[0] ) {
+            $id = $1;
+        }
+    }
+    return PE_OK unless $id;
+    my $session = $self->getRefreshToken($id);
+    return PE_OK unless $session;
+    my $rp     = $self->getRP($session->data->{client_id});
+    my $bck = $self->rpOptions->{$rp}->{oidcRPMetaDataOptionsLogoutType};
+    if ( $bck and $bck eq 'back' ) {
+        if ( my $err = $self->_backChannelLogout( $req, $rp, $session->data ) ) {
+            return $err unless $msg;
+        }
+    }
+    $session->remove;
+    return PE_OK;
+}
+
 1;
--- a/usr/share/perl5/Lemonldap/NG/Portal/Lib/OpenIDConnect.pm
+++ b/usr/share/perl5/Lemonldap/NG/Portal/Lib/OpenIDConnect.pm
@@ -1288,9 +1288,7 @@ sub makeJWT {
     return undef if ( $h != PE_OK );
 
     # Get signature algorithm
-    my $alg = $self->rpOptions->{$rp}->{oidcRPMetaDataOptionsAccessTokenSignAlg}
-      || ( $self->conf->{oidcServiceKeyTypeSig} eq 'EC' ? 'ES256' : 'RS256' );
-    $self->logger->debug("Access Token signature algorithm: $alg");
+    my $alg = $self->getSignAlg($rp);
 
     my $jwt = $self->createJWT( $access_token_payload, $alg, $rp, "at+JWT" );
 
@@ -1472,6 +1470,14 @@ sub getOpenIDConnectSession {
     return $oidcSession;
 }
 
+sub getSignAlg {
+    my ( $self, $rp ) = @_;
+    my $alg = $self->rpOptions->{$rp}->{oidcRPMetaDataOptionsAccessTokenSignAlg}
+      || ( $self->conf->{oidcServiceKeyTypeSig} eq 'EC' ? 'ES256' : 'RS256' );
+    $self->logger->debug("Access Token signature algorithm: $alg");
+    return $alg;
+}
+
 # Store information in state database and return
 # corresponding session_id
 # @return State Session ID
--- a/usr/share/perl5/Lemonldap/NG/Portal/Main/Run.pm
+++ b/usr/share/perl5/Lemonldap/NG/Portal/Main/Run.pm
@@ -322,13 +322,14 @@ sub logout {
 
 sub unauthLogout {
     my ( $self, $req ) = @_;
-    $self->_unauthLogout($req);
-    return $self->do( $req, [ 'controlUrl', sub { PE_LOGOUT_OK } ] );
+    my $err = $self->_unauthLogout($req);
+    return $self->do( $req,
+        [ 'controlUrl', sub { $err > 0 ? $err : PE_LOGOUT_OK } ] );
 }
 
 sub _unauthLogout {
     my ( $self, $req ) = @_;
-    $self->processHook( $req, 'unAuthLogout' );
+    my $err = $self->processHook( $req, 'unAuthLogout' );
     $self->logger->debug('Unauthenticated logout request');
     $self->logger->debug('Cleaning pdata');
     $self->logger->debug("Removing $self->{conf}->{cookieName} cookie");
@@ -342,6 +343,7 @@ sub _unauthLogout {
             value   => 0
         )
     );
+    return $err;
 }
 
 # RUNNING METHODS
