--- a/usr/share/perl5/Lemonldap/NG/Portal/Issuer/OpenIDConnect.pm
+++ b/usr/share/perl5/Lemonldap/NG/Portal/Issuer/OpenIDConnect.pm
@@ -2927,6 +2927,10 @@ sub metadata {
 sub exportRequestParameters {
     my ( $self, $req ) = @_;
 
+    unless ( $self->ssoMatch($req) ) {
+        return PE_OK;
+    }
+
     if ( my $p = $req->param('prompt') ) {
         if ( $p eq 'none' ) {
             return $self->_unauthPromptNone($req);
