--- a/usr/share/perl5/Lemonldap/NG/Portal/Lib/OpenIDConnect.pm
+++ b/usr/share/perl5/Lemonldap/NG/Portal/Lib/OpenIDConnect.pm
@@ -490,7 +490,7 @@ sub buildAuthorizationCodeAuthnRequest {
                 iat => time,
                 %$authorize_request_params,
             },
-            $self->opOptions->{op}
+            $self->opOptions->{$op}
               ->{oidcOPMetaDataOptionsAuthnEndpointAuthSigAlg} || 'RS256',
             $op
         );
