diff --git a/usr/share/perl5/Lemonldap/NG/Portal/Auth/_WebForm.pm b/usr/share/perl5/Lemonldap/NG/Portal/Auth/_WebForm.pm
index a99c54442..f5e63043f 100644
--- a/usr/share/perl5/Lemonldap/NG/Portal/Auth/_WebForm.pm
+++ b/usr/share/perl5/Lemonldap/NG/Portal/Auth/_WebForm.pm
@@ -51,7 +51,9 @@ sub init {
     );
 
     return 0 unless $self->captcha;
-    unless ( $self->{conf}->{captcha_login_enabled} ) {
+    unless ($self->{conf}->{captcha_login_enabled}
+        and $self->{conf}->{captcha_login_enabled} eq '0' )
+    {
         $self->ott( $self->p->loadModule('::Lib::OneTimeToken') ) or return 0;
         $self->ott->timeout( $self->conf->{formTimeout} );
     }
