--- a/usr/share/perl5/Lemonldap/NG/Portal/Issuer/OpenIDConnect.pm
+++ b/usr/share/perl5/Lemonldap/NG/Portal/Issuer/OpenIDConnect.pm
@@ -246,6 +246,8 @@ sub loadRPs {
 
 sub ssoMatch {
     my ( $self, $req ) = @_;
+    my $lh = $req->param('login_hint');
+    $req->data->{suggestedLogin} ||= $lh if $lh;
     return ( $req->uri =~ $self->ssoMatchUrl ? 1 : 0 );
 }
 
--- a/usr/share/perl5/Lemonldap/NG/Portal/Lib/OpenIDConnect.pm
+++ b/usr/share/perl5/Lemonldap/NG/Portal/Lib/OpenIDConnect.pm
@@ -429,6 +429,7 @@ sub buildAuthorizationCodeAuthnRequest {
     my $max_age       = $self->opOptions->{$op}->{oidcOPMetaDataOptionsMaxAge};
     my $ui_locales = $self->opOptions->{$op}->{oidcOPMetaDataOptionsUiLocales};
     my $acr_values = $self->opOptions->{$op}->{oidcOPMetaDataOptionsAcrValues};
+    my $login_hint = $req->data->{suggestedLogin};
 
     my $authorize_request_oauth2_params = {
         response_type => $response_type,
@@ -437,6 +438,7 @@ sub buildAuthorizationCodeAuthnRequest {
         redirect_uri  => $redirect_uri,
         ( defined $state ? ( state => $state ) : () ),
         ( defined $nonce ? ( nonce => $nonce ) : () ),
+        ( defined $login_hint ? ( login_hint => $login_hint) : () ),
     };
     my $authorize_request_params = {
         %$authorize_request_oauth2_params,
--- a/usr/share/perl5/Lemonldap/NG/Portal/Main/Display.pm
+++ b/usr/share/perl5/Lemonldap/NG/Portal/Main/Display.pm
@@ -195,6 +195,7 @@ sub display {
             ASK_STAYCONNECTED => $req->param('stayconnected') || 0,
             CONFIRMKEY        => $self->stamp(),
             LIST              => $req->data->{list} || [],
+            LOGIN_HINT        => $req->data->{suggestedLogin},
             (
                 $req->data->{customScript}
                 ? ( CUSTOM_SCRIPT => $req->data->{customScript} )
@@ -412,7 +413,7 @@ sub display {
     # 3 Authentication has been refused OR first access
     else {
         $skinfile = 'login';
-        my $login = $req->user;
+        my $login = $req->user || $req->data->{suggestedLogin};
         %templateParams = (
             $self->getErrorTplParams($req),
             AUTH_URL              => $req->{data}->{_url},
--- a/usr/share/lemonldap-ng/portal/templates/bootstrap/idpchoice.tpl
+++ b/usr/share/lemonldap-ng/portal/templates/bootstrap/idpchoice.tpl
@@ -15,6 +15,9 @@
     <TMPL_IF NAME="CONFIRMKEY">
       <input type="hidden" id="confirm" name="confirm" value="<TMPL_VAR NAME="CONFIRMKEY">" />
     </TMPL_IF>
+    <TMPL_IF NAME="LOGIN_HINT">
+      <input type="hidden" name="login_hint" value="<TMPL_VAR NAME="LOGIN_HINT">" />
+    </TMPL_IF>
     <input type="hidden" name="skin" value="<TMPL_VAR NAME="SKIN">" />
 
     <div class="card border-info">
