--- a/usr/share/perl5/Lemonldap/NG/Handler/PSGI/Try.pm
+++ b/usr/share/perl5/Lemonldap/NG/Handler/PSGI/Try.pm
@@ -92,7 +92,18 @@ sub _run {
                 $req->userData( $self->api->data );
                 $req->respHeaders( $res->[1] );
             }
-            elsif ( $res->[0] != 403 and not $req->data->{noTry} ) {
+            elsif ( ( (
+                            $res->[0] != 403
+                        and $res->[0] != 503
+                        and $res->[0] != 302
+                    )
+                    or $req->path_info =~ m#^/+lmerror/#i
+                    or ( $res->[0] == 302
+                        and not grep( m#/lmerror/503#, @{ $res->[1] } ) )
+                )
+                and not $req->data->{noTry}
+              )
+            {
 
                 # Unset headers (handler adds a Location header)
                 $self->logger->debug(
