--- a/usr/share/perl5/Lemonldap/NG/Portal/Main/Process.pm
+++ b/usr/share/perl5/Lemonldap/NG/Portal/Main/Process.pm
@@ -238,7 +238,7 @@ sub checkUnauthLogout {
                 value   => 0
             )
         );
-        $req->steps( [ sub { PE_LOGOUT_OK } ] );
+        $req->steps( [ 'controlUrl', sub { PE_LOGOUT_OK } ] );
     }
     return PE_OK;
 }
--- a/usr/share/perl5/Lemonldap/NG/Portal/Main/Run.pm
+++ b/usr/share/perl5/Lemonldap/NG/Portal/Main/Run.pm
@@ -283,7 +283,7 @@ sub unauthLogout {
             value   => 0
         )
     );
-    return $self->do( $req, [ sub { PE_LOGOUT_OK } ] );
+    return $self->do( $req, [ 'controlUrl', sub { PE_LOGOUT_OK } ] );
 }
 
 # RUNNING METHODS
