--- a/usr/share/perl5/Lemonldap/NG/Portal/Issuer/OpenIDConnect.pm
+++ b/usr/share/perl5/Lemonldap/NG/Portal/Issuer/OpenIDConnect.pm
@@ -2329,20 +2329,27 @@ sub introspection {
         my $apacheSession =
           $self->_getSessionFromAccessTokenData( $oidcSession->{data} );
         if ($apacheSession) {
-
-            $response->{active} = JSON::true;
-
-        # The ID attribute we choose is the one of the calling webservice,
-        # which might be different from the OIDC client the token was issued to.
-            $response->{sub} =
-              $self->getUserIDForRP( $req, $rp, $apacheSession->data );
-            $response->{scope} = $oidcSession->{data}->{scope}
-              if $oidcSession->{data}->{scope};
-            $response->{client_id} = $oidcSession->{data}->{client_id}
-              if $oidcSession->{data}->{client_id};
-            $response->{iss} = $self->get_issuer($req);
-            $response->{exp} =
-              $oidcSession->{data}->{_utime} + $self->conf->{timeout};
+            $response = {
+                active    => JSON::true,
+                aud       => $self->getAudiences( $oidcSession->data->{rp} ),
+                client_id => $oidcSession->{data}->{client_id},
+                exp   => int($oidcSession->data->{_utime} + $self->conf->{timeout}),
+                iat   => int($oidcSession->data->{iat}),
+                iss   => $self->get_issuer($req),
+                nbf   => int($oidcSession->data->{nbf}),
+                scope => $oidcSession->{data}->{scope},
+                token_type => 'bearer',
+
+                # The ID attribute we choose is the one of the calling
+                # webservice, which might be different from the OIDC client the
+                # token was issued to.
+                sub => $self->getUserIDForRP( $req, $rp, $apacheSession->data ),
+            };
+            foreach ( keys %$response ) {
+                delete $response->{$_}
+                  unless defined $response->{$_}
+                  and $response->{$_} || length( $response->{$_} ) > 0;
+            }
         }
         else {
             $self->logger->warn("Count not find session tied to Access Token");
