--- a/usr/share/perl5/Lemonldap/NG/Portal/Issuer/OpenIDConnect.pm
+++ b/usr/share/perl5/Lemonldap/NG/Portal/Issuer/OpenIDConnect.pm
@@ -2651,7 +2651,9 @@ sub logout {
                                 "OIDC back channel: unable to unlog"
                               . " $userId from $rp: "
                               . $resp->message );
-                        $self->logger->debug( $resp->content );
+                        $self->logger->debug("Logout token: $jwt");
+                        $self->logger->debug( 'Upstream status: ' . $resp->status_line );
+                        $self->logger->debug( 'Upstream response: ' . ( $resp->content // '') );
                         $code = PE_SLO_ERROR;
                     }
                     else {
