--- a/usr/share/perl5/Lemonldap/NG/Portal/Issuer/OpenIDConnect.pm
+++ b/usr/share/perl5/Lemonldap/NG/Portal/Issuer/OpenIDConnect.pm
@@ -2032,6 +2032,7 @@ sub _handleRefreshTokenGrant {
         }
 
         # Lookup attributes and macros for user
+        $self->p->processHook( $req, 'oidcGotOfflineRefreshData', $rp, $refreshSession->data );
         $self->getAttributesForUser( $req, $refreshSession )
           or return $self->sendOIDCError( $req, 'invalid_grant', 400 );
 
