--- a/usr/share/perl5/Lemonldap/NG/Portal/Lib/OpenIDConnect.pm
+++ b/usr/share/perl5/Lemonldap/NG/Portal/Lib/OpenIDConnect.pm
@@ -1240,7 +1240,7 @@ sub makeJWT {
       || ( $self->conf->{oidcServiceKeyTypeSig} eq 'EC' ? 'ES256' : 'RS256' );
     $self->logger->debug("Access Token signature algorithm: $alg");
 
-    my $jwt = $self->createJWT( $access_token_payload, $alg, $rp, "at+JWT" );
+    my $jwt = $self->createJWT( $access_token_payload, $alg, $rp );
 
     return $jwt;
 }
