--- a/usr/share/perl5/Lemonldap/NG/Portal/Issuer/OpenIDConnect.pm
+++ b/usr/share/perl5/Lemonldap/NG/Portal/Issuer/OpenIDConnect.pm
@@ -2704,8 +2704,13 @@ sub logout {
 sub metadata {
     my ( $self, $req ) = @_;
     $req->data->{dropCsp} = 1 if $self->conf->{oidcDropCspHeaders};
+    my %args;
+    if ( $self->conf->{oidcServiceMetadataTtl} ) {
+        $args{ttl}  = $self->conf->{oidcServiceMetadataTtl};
+        $args{date} = $self->conf->{cfgDate};
+    }
     return $self->p->sendJSONresponse( $req,
-        $self->metadataDoc( $self->get_issuer($req) ) );
+        $self->metadataDoc( $self->get_issuer($req) ), %args );
 }
 
 # Store request parameters in %ENV
