--- a/usr/share/perl5/Lemonldap/NG/Portal/Main/Display.pm
+++ b/usr/share/perl5/Lemonldap/NG/Portal/Main/Display.pm
@@ -596,12 +596,16 @@ sub display {
             if ( $self->conf->{authentication} eq 'Choice'
                 and my $authLoop = $self->_buildAuthLoop($req) )
             {
+                my $tab = $req->param("tab");
+                if ($tab) {
+                    $tab = undef unless grep { $_->{key} eq $tab } @$authLoop;
+                }
                 %templateParams = (
                     %templateParams,
                     AUTH_LOOP            => $authLoop,
                     CHOICE_PARAM         => $self->conf->{authChoiceParam},
                     CHOICE_VALUE         => $req->data->{_authChoice},
-                    DISPLAY_TAB          => scalar( $req->param("tab") ),
+                    DISPLAY_TAB          => $tab,
                     DISPLAY_FORM         => 0,
                     DISPLAY_OPENID_FORM  => 0,
                     DISPLAY_YUBIKEY_FORM => 0,
